src/game_api/strings.cpp

```diff
@@ -0,0 +1,121 @@
+#include "strings.hpp"
+#include "entity.hpp"
+#include "fix_entity_descriptions.hpp"
+
+#include "detours.h"
+#include <cassert>
+
+static STRINGID wrong_stringid = 0;
+std::map<STRINGID, std::u16string> custom_strings;
+std::map<uint32_t, STRINGID> string_hashes = {
+    {0xd41f49aa, 0},
+    {0xcd92c15d, 1}
+    // TODO: make script to dump all automatically
+};
+
+using OnShopItemNameFormatFun = void(Entity*, char16_t*);
+OnShopItemNameFormatFun* g_on_shopnameformat_trampoline{nullptr};
+void on_shopitemnameformat(Entity* item, char16_t* buffer)
+{
+    const STRINGID items_stringid = item->type->description;
+    if (items_stringid >= wrong_stringid)
+    {
+        const STRINGID buy_stringid = 1340; // id of the "Buy %s" text //TODO: replace with hash_to_stringid(0x21683743); when dumping script is done
+        constexpr auto buffer_size = 30;    //guess the buffer size, for now
+
+        swprintf_s((wchar_t*)buffer, buffer_size, (wchar_t*)get_string(buy_stringid), get_string(items_stringid));
+        return;
+    }
+
+    g_on_shopnameformat_trampoline(item, buffer);
+}
+
+void strings_init()
+{
+    if (wrong_stringid == 0)
+    {
+        //get wrong stringid from bordertile
+        wrong_stringid = get_type(1)->description;
+    }
+    fix_entity_descriptions();
+
+    auto addr_insta = get_address("get_shopitem_name");
+    g_on_shopnameformat_trampoline = (OnShopItemNameFormatFun*)addr_insta;
+    DetourTransactionBegin();
+    DetourUpdateThread(GetCurrentThread());
+
+    DetourAttach((void**)&g_on_shopnameformat_trampoline, &on_shopitemnameformat);
+    const LONG error = DetourTransactionCommit();
+    if (error != NO_ERROR)
+    {
+        DEBUG("Failed hooking get_shopitem_name: {}\n", error);
+    }
+}
+
+const char16_t** get_strings_table()
+{
+    static const char16_t** string_table = (const char16_t**)get_address("string_table");
+    return string_table;
+}
+
+STRINGID hash_to_stringid(uint32_t hash)
+{
+    if (string_hashes.find(hash) != string_hashes.end())
```

Comment:
`auto it = string_hashes.find(hash);` and then `return it->second` is a slight optimization here